barretenberg/cpp/pil/avm/alu_chip.pil

```diff
@@ -0,0 +1,214 @@
+    // number of registers and complexity of the range constraints.
+    // TODO: Determine the best slice size allocation. Note that we might choose different
+    // slice sizes for the different types. At this stage, it is still unknown how this
+    // allocation affects the performance.
+    //
+    // Preliminary choice: We use one u8 register to support u8 operations and up to 8 u16
+    // registers for the other types.
+
+
+    // ============= Helper polynomial terms ============================
+    // These are intermediate polynomial terms which are not commited but
+    // serves to express commited polynomials in a more concise way.
+
+    // 16-bit slice partial sums
+    pol sum_16 =  alu_u16_r0;
+    pol sum_32 =  sum_16     + alu_u16_r1 * 2**16;
+    pol sum_64 =  sum_32     + alu_u16_r2 * 2**32 + alu_u16_r3 * 2**48; 
+    pol sum_96 =  sum_64     + alu_u16_r4 * 2**64 + alu_u16_r5 * 2**80;
+    pol sum_128 = sum_96     + alu_u16_r6 * 2**96 + alu_u16_r7 * 2**112;
+
+    // ========= ADDITION Operation Constraints ===============================
+
+    // ff addition
+    #[SUBOP_ADDITION_FF]
+    alu_ff_tag * alu_op_add * (alu_ia + alu_ib - alu_ic) = 0;
+
+    // u8 addition
+    alu_u8_tag * alu_op_add * (alu_u8_r0 + alu_cf * 2**8 - alu_ia - alu_ib) = 0;
+    alu_u8_tag * alu_op_add * (alu_u8_r0 - alu_ic) = 0;
+
+    // u16 addition
+    alu_u16_tag * alu_op_add * (sum_16 + alu_cf * 2**16  - alu_ia - alu_ib) = 0;
+    alu_u16_tag * alu_op_add * (sum_16 - alu_ic) = 0;
+
+    // u32 addition
+    alu_u32_tag * alu_op_add * (sum_32 + alu_cf * 2**32 - alu_ia - alu_ib) = 0;
+    alu_u32_tag * alu_op_add * (sum_32 - alu_ic) = 0;
+
+    // u64 addition
+    alu_u64_tag * alu_op_add * (sum_64 + alu_cf * 2**64 - alu_ia - alu_ib) = 0;
+    alu_u64_tag * alu_op_add * (sum_64 - alu_ic) = 0;
+
+    // u128 addition
+    alu_u128_tag * alu_op_add * (sum_128 + alu_cf * 2**128 - alu_ia - alu_ib) = 0;
+    alu_u128_tag * alu_op_add * (sum_128 - alu_ic) = 0;
+
+    // ========= SUBTRACTION Operation Constraints ===============================
+
+    // a - b = c <==> c + b = a (mod 2^k)
+    // Same constraints as for addition but we swap alu_ia with alu_ic
+
+    // ff subtraction
+    #[SUBOP_SUBTRACTION_FF]
+    alu_ff_tag * alu_op_sub * (alu_ia - alu_ib - alu_ic) = 0;
+
+    // u8 subtraction
+    alu_u8_tag * alu_op_sub * (alu_u8_r0 + alu_cf * 2**8 - alu_ic - alu_ib) = 0;
+    alu_u8_tag * alu_op_sub * (alu_u8_r0 - alu_ia) = 0;
+
+    // u16 subtraction
+    alu_u16_tag * alu_op_sub * (sum_16 + alu_cf * 2**16 - alu_ic - alu_ib) = 0;
+    alu_u16_tag * alu_op_sub * (sum_16 - alu_ia) = 0;
+
+    // u32 subtraction
+    alu_u32_tag * alu_op_sub * (sum_32 + alu_cf * 2**32 - alu_ic - alu_ib) = 0;
+    alu_u32_tag * alu_op_sub * (sum_32 - alu_ia) = 0;
+
+    // u64 subtraction
+    alu_u64_tag * alu_op_sub * (sum_64 + alu_cf * 2**64 - alu_ic - alu_ib) = 0;
+    alu_u64_tag * alu_op_sub * (sum_64 - alu_ia) = 0;
+
+    // u128 subtraction
```

Comment:
This is brilliant @zac-williamson ! After some thoughts, I am convinced this is working because arithmetic modulo 2^l maps the same arithmetic modulo 2^k where k<l.
I will re-implement addition and subtraction relations with these ideas.